nixos/modules/system/boot/systemd/initrd.nix

```diff
@@ -0,0 +1,379 @@
+        { object = "${pkgs.glibc}/lib/libnss_files.so"; }
+
+        { object = config.environment.etc.os-release.source; symlink = "/etc/initrd-release"; }
+        { object = config.environment.etc.os-release.source; symlink = "/etc/os-release"; }
+        { object = fstab; symlink = "/etc/fstab"; }
+        {
+          object = "${modulesClosure}/lib/modules";
+          symlink = "/lib/modules";
+        }
+        {
+          symlink = "/etc/modules-load.d/nixos.conf";
+          object = pkgs.writeText "nixos.conf"
+            (lib.concatStringsSep "\n" config.boot.initrd.kernelModules);
+        }
+        {
+          object = builtins.toFile "passwd" "root:x:0:0:System Administrator:/root:/bin/bash";
+          symlink = "/etc/passwd";
+        }
+        {
+          object = builtins.toFile "shadow" "root:${config.boot.initrd.systemd.emergencyHashedPassword}:::::::";
+          symlink = "/etc/shadow";
+        }
+        { object = "${initrdBinEnv}/bin"; symlink = "/bin"; }
+        { object = "${initrdBinEnv}/sbin"; symlink = "/sbin"; }
+        { object = builtins.toFile "sysctl.conf" "kernel.modprobe = /sbin/modprobe"; symlink = "/etc/sysctl.d/nixos.conf"; }
+      ];
+
+      targets.initrd.aliases = ["default.target"];
+      units =
+           mapAttrs' (n: v: nameValuePair "${n}.path"    (pathToUnit          n v)) cfg.paths
+        // mapAttrs' (n: v: nameValuePair "${n}.service" (initrdServiceToUnit n v)) cfg.services
+        // mapAttrs' (n: v: nameValuePair "${n}.slice"   (sliceToUnit         n v)) cfg.slices
+        // mapAttrs' (n: v: nameValuePair "${n}.socket"  (socketToUnit        n v)) cfg.sockets
+        // mapAttrs' (n: v: nameValuePair "${n}.target"  (targetToUnit        n v)) cfg.targets
+        // mapAttrs' (n: v: nameValuePair "${n}.timer"   (timerToUnit         n v)) cfg.timers
+        // listToAttrs (map
+                     (v: let n = escapeSystemdPath v.where;
+                         in nameValuePair "${n}.mount" (mountToUnit n v)) cfg.mounts)
+        // listToAttrs (map
+                     (v: let n = escapeSystemdPath v.where;
+                         in nameValuePair "${n}.automount" (automountToUnit n v)) cfg.automounts);
+
+      # The unit in /run/systemd/generator shadows the unit in
+      # /etc/systemd/system, but will still apply drop-ins from
+      # /etc/systemd/system/foo.service.d/
+      #
+      # We need IgnoreOnIsolate, otherwise the Requires dependency of
+      # a mount unit on its makefs unit causes it to be unmounted when
+      # we isolate for switch-root. Use a dummy package so that
+      # generateUnits will generate drop-ins instead of unit files.
+      packages = [(pkgs.runCommand "dummy" {} ''
+        mkdir -p $out/etc/systemd/system
+        touch $out/etc/systemd/system/systemd-{makefs,growfs}@.service
+      '')];
+      services."systemd-makefs@".unitConfig.IgnoreOnIsolate = true;
+      services."systemd-growfs@".unitConfig.IgnoreOnIsolate = true;
```

Comment:
Odd. Don't other distros with systemd-in-initrd run into the same problem? What do they use? Maybe we can link from here.

Reply:
@flokli I'm not sure that other distros actually use `systemd-makefs` or `systemd-growfs` in initrd. [For instance, the fedora guide for installing a fedora image on an sd card for raspberry pi](https://docs.fedoraproject.org/en-US/quick-docs/raspberry-pi/#installing-fedora-on-a-raspberry-pi-for-linux-users_rpi) tells you to resize the root fs manually before booting the sd image on the pi.

Reply:
Hmmh, so we run the fstab generator in initrd, right?

It generates a unit file like this (as per https://github.com/systemd/systemd/blob/main/src/shared/generator.c#L466-L486 and https://github.com/systemd/systemd/blob/main/src/shared/generator.c#L532-L543):
```
        fprintf(f,
                "# Automatically generated by %s\n\n"
                "[Unit]\n"
                "Description=Make Swap on %%f\n"
                "Documentation=man:systemd-mkswap@.service(8)\n"
                "DefaultDependencies=no\n"
                "BindsTo=%%i.device\n"
                "Conflicts=shutdown.target\n"
                "After=%%i.device\n"
                "Before=shutdown.target %s\n"
                "\n"
                "[Service]\n"
                "Type=oneshot\n"
                "RemainAfterExit=yes\n"
                "ExecStart="SYSTEMD_MAKEFS_PATH " swap %s\n"
                "TimeoutSec=0\n",
                program_invocation_short_name,
                where_unit,
                escaped);
```

> # We need IgnoreOnIsolate, otherwise the Requires dependency of
> # a mount unit on its makefs unit causes it to be unmounted when
> # we isolate for switch-root. …

I don't think it's a `Requires` on the mount unit:

```
# Automatically generated by systemd-fstab-generator

[Unit]
Documentation=man:fstab(5) man:systemd-fstab-generator(8)
SourcePath=/etc/fstab
Before=local-fs.target
After=blockdev@dev-mapper-root.target

[Mount]
Where=/
What=/dev/mapper/root
Type=btrfs
Options=…
```

Could it be because we don't use `systemctl switch-root`?:

```
       switch-root ROOT [INIT]
           Switches to a different root directory and executes a new system manager process below it. This is intended for usage in initial RAM disks ("initrd"), and
           will transition from the initrd's system manager process (a.k.a. "init" process) to the main system manager process which is loaded from the actual host
           volume. This call takes two arguments: the directory that is to become the new root directory, and the path to the new system manager binary below it to
           execute as PID 1. If the latter is omitted or the empty string, a systemd binary will automatically be searched for and used as init. If the system manager
           path is omitted, equal to the empty string or identical to the path to the systemd binary, the state of the initrd's system manager process is passed to the
           main system manager, which allows later introspection of the state of the services involved in the initrd boot phase.
```

Reply:
I looked at the vm test logs:

```
machine # [  OK  ] Finished Cleanup udev Database.
machine # [  OK  ] Reached target Switch Root.
machine #          Starting Switch Root...
machine # [    1.881830] systemd-journald[73]: Received SIGTERM from PID 1 (systemd).
machine # 
machine # <<< NixOS Stage 2 >>>
machine # 
machine # [    1.967187] EXT4-fs (vda): re-mounted. Opts: (null). Quota mode: none.
machine # [    1.970678] booting system configuration /nix/store/c2birzmvm77j79lh5njrz18z2nrjg50i-nixos-system-machine-22.05.git.82102fc37da
machine # running activation script...
machine # setting up /etc...
machine # starting systemd...
machine # [    4.579760] systemd[1]: Inserted module 'autofs4'
machine # [    4.603075] systemd[1]: systemd 249.7 running in system mode (+PAM +AUDIT -SELINUX +APPARMOR +IMA +SMACK +SECCOMP +GCRYPT -GNUTLS +OPENSSL +ACL +BLKID +CURL +ELFUTILS +FIDO2 +IDN2 -IDN +IPTC +KMOD +LIBCRYPTSETUP +LIBFDISK +PCRE2 -PWQUALITY -P11KIT -QRENCODE +BZIP2 +LZ4 +XZ +ZLIB +ZSTD -XKBCOMMON +UTMP -SYSVINIT default-hierarchy=unified)
```

It looks like `initrd-switch-root.service` fails. It'd run `systemctl --no-block switch-root /sysroot`, but we don't use /sysroot inside the initrd, but /mnt-root. We then for some reason still run stage 2, which does exec systemd, instead of using `systemctl switch-root` - which might explain the `IgnoreOnIsolate` hacks we need to add currently.

Reply:
It's called `/sysroot` on the systemd-based stage-1 and `/mnt-root` on the bash one

Reply:
@flokli If I force stage 1 into emergency mode and run `systemctl show sysroot.mount`, I see `Requires=dev-vda.device systemd-makefs@dev-vda.service system.slice`, meaning it does have a `Requires=` dependency on the makefs service. I remember digging through the fstab-generator code for quite a while before I figured out how it worked (the unit files themselves shown by `systemctl cat ...` don't make it clear either), but unfortunately I don't remember the cause I found for this `Requires=` dependency. Regardless, it is there, so `sysroot` gets unmounted without this `IgnoreOnIsolate` stuff.

Reply:
Ah, something creates a `/run/systemd/generator/sysroot.mount.requires` directory with a symlink to the makefs unit.

Reply:
I guess it's fine to initially keep this, but we should add a comment, and maybe revisit this while addressing https://github.com/NixOS/nixpkgs/issues/165649.

Reply:
There is a comment, and I'm not sure what this has to do with #165649?

Reply:
Checking why we have to set `IgnoreOnIsolate` here.

I assume if we stop switching to stage2, but directly to the other systemd (via `systemctl switch-root`) things could work.

Reply:
No, that won't have any effect on this. The problem is just that stage 1 will do `systemctl --no-block isolate initrd-switch-root.target`, which goes on a killing spree to stop all units. It'll do this whether we have systemd-to-systemd handoff or not. Once the makefs unit gets stopped, because of the Requires= dependency, /sysroot will get unmounted. So we need `IgnoreOnIsolate` so that the makefs unit never gets stopped

Reply:
OK. I feel like this smells like something upstream could handle more gracefully, and we might want to have a talk with them about it at some time, but nothing that should block this ;-)